packages/xo-server-load-balancer/src/plan.js

```diff
@@ -276,6 +278,189 @@ export default class Plan {
+    if (allHosts.length <= 1) {
+      return
+    }
+    const idToHost = keyBy(allHosts, 'id')
+    const allVms = filter(this._getAllRunningVms(), vm => vm.$container in idToHost)
+    const hostList = this._getVCPUHosts(allHosts, allVms)
+    const idealVcpuPerCpuRatio =
+      hostList.reduce((sum, host) => sum + host.vcpuCount, 0) / hostList.reduce((sum, host) => sum + host.cpuCount, 0)
+
+    debugVcpuBalancing('Try to apply VCPU prepositionning.')
+    debugVcpuBalancing(`VCPU count per host: ${inspect(hostList, { depth: null })}`)
+    debugVcpuBalancing(`Average vCPU per CPU : ${idealVcpuPerCpuRatio}`)
+
+    const vmsAverages = await this._getVmsAverages(allVms, idToHost)
+    const { averages: hostsAverages } = await this._getHostStatsAverages({ hosts: allHosts })
+
+    // 1. Find source host from which to migrate.
+    const sources = sortBy(
+      filter(hostList, host => (host.vcpuCount - 1) / host.cpuCount >= idealVcpuPerCpuRatio),
+      [
+        host => -host.vcpuCount / host.cpuCount,
+        // Find host with the most memory used
+        host => hostsAverages[host.id].memoryFree,
+      ]
+    )
+    debugVcpuBalancing(`Sources : ${inspect(sources, { depth: null })}`)
+
+    for (const sourceHost of sources) {
+      let deltaSource = sourceHost.vcpuCount - sourceHost.cpuCount * idealVcpuPerCpuRatio
+      if (deltaSource < 1) {
+        continue
+      }
+
+      // 2. Find destination host.
+      const destinations = sortBy(
+        filter(hostList, host => host.id !== sourceHost.id && host.vcpuCount < host.cpuCount * idealVcpuPerCpuRatio),
+        [
+          // trying to avoid migrations between pools
+          host => host.poolId === sourceHost.poolId,
+          host => host.vcpuCount / host.cpuCount,
+          host => -hostsAverages[host.id].memoryFree,
+        ]
+      )
+      debugVcpuBalancing(`Destinations : ${inspect(destinations, { depth: null })}`)
+
+      if (!destinations.length) {
+        continue // Cannot find a valid destination.
+      }
+
+      // Build VM list to migrate.
+      const sourceVms = Object.values(sourceHost.vms)
+
+      // eslint-disable-next-line no-labels
+      destinationLoop: for (const destinationHost of destinations) {
+        debugVcpuBalancing(`Host candidate: ${sourceHost.id} -> ${destinationHost.id}`)
+
+        // calculating how many vcpu source should give and how many destination should accept
+        let deltaDestination = destinationHost.vcpuCount - destinationHost.cpuCount * idealVcpuPerCpuRatio
+
+        if (
+          deltaDestination >= 0 ||
+          hostsAverages[destinationHost.id].cpu >= this._thresholds.cpu.low ||
+          hostsAverages[destinationHost.id].memoryFree <= this._thresholds.memoryFree.low
+        ) {
+          continue
+        }
+
+        // avoiding to migrate too much vCPUs for source or destination
+        let delta =
+          Math.abs(deltaSource) > Math.abs(deltaDestination)
```

Comment:
Are you sure we must use abs here?

Reply:
We could probably avoid it as we know one is positive and the other one is negative